tests/testthat/test-write_dataset_json.R

```diff
@@ -144,3 +144,97 @@ test_that("write_dataset_json errors are thrown properly", {
     "Folder supplied to `file` does not exist"
   )
 })
+
+test_that("datetime and times write out properly", {
+  df_name <- "adsl"
+  orig_df <- readRDS(testthat::test_path("testdata", "adsl_time_test.Rds"))
+  df_metadata <- readRDS(testthat::test_path("testdata", "adsl_time_test_meta.Rds"))
+
+  # create dataset json object
+  ds_json <- dataset_json(
+    orig_df,
+    file_oid = "www.cdisc.org/StudyMSGv1/1/Define-XML_2.1.0/2024-11-11/adsl",
+    last_modified = "2022-04-16T20:09:03",
+    originator = "CDISC ADaM MSG Team",
+    sys = "SAS on X64_10PRO",
+    sys_version = "9.0401M7",
+    study = "TDF_ADaM.ADaMIG.1.1",
+    metadata_version = "MDV.TDF_ADaM.ADaMIG.1.1",
+    metadata_ref = "define.xml",
+    item_oid = "IG.ADSL",
+    name = "ADSL",
+    dataset_label = "Subject-Level Analysis Dataset",
+    columns = df_metadata
+  )
+
+  # Write JSON
+  adsl_json_output <- write_dataset_json(ds_json)
+  adsl_json_input <- read_dataset_json(adsl_json_output)
+
+  # The ignore_attr option isn't working here.
+  # Period objects (i.e. times)
+  x <- orig_df$VISIT1TM
+  y <- adsl_json_input$VISIT1TM
+  attr(y, 'label') <- NULL
+  expect_equal(x, y)
+
+  # Datetimes
+  x <- orig_df$VIST1DTM
+  y <- adsl_json_input$VIST1DTM
+  attr(y, 'label') <- NULL
+  expect_equal(x, y)
+
+  # Dates
+  x <- orig_df$VISIT1DT
+  y <- adsl_json_input$VISIT1DT
+  attr(x, 'format.sas') <- NULL
+  expect_equal(x, y)
+
+
+
+})
+
+make_ds_json <- function(dat, meta) {
+  dataset_json(
+    dat,
+    file_oid = "www.cdisc.org/StudyMSGv1/1/Define-XML_2.1.0/2024-11-11/adsl",
+    last_modified = "2022-04-16T20:09:03",
+    originator = "CDISC ADaM MSG Team",
+    sys = "SAS on X64_10PRO",
+    sys_version = "9.0401M7",
+    study = "TDF_ADaM.ADaMIG.1.1",
+    metadata_version = "MDV.TDF_ADaM.ADaMIG.1.1",
+    metadata_ref = "define.xml",
+    item_oid = "IG.ADSL",
+    name = "ADSL",
+    dataset_label = "Subject-Level Analysis Dataset",
+    columns = meta
+  )
+}
+
+test_that("Writing errors trigger", {
+  orig_df <- readRDS(testthat::test_path("testdata", "adsl_time_test.Rds"))
+  df_metadata <- readRDS(testthat::test_path("testdata", "adsl_time_test_meta.Rds"))
+
+  # fails for POSIXct
```

Comment:
Do we explain why we require POSIXlt?  I haven't gotten to the vignettes yet.  Sorry if this is already addressed.

Reply:
Yeah:

> -   Numeric date times will be converted to the base R type of `POSIXlt`
>    - `POSIXlt` is used as opposed to `POSIXct` because we require that datetimes are set to the UTC timezone. CDISC dates are generally not timezone qualified, and unless a timezone is explicitly specified systems may default to the user's current timezone. As such, we've introduced a hard requirement that datetimes are anchored to UTC. If the datetime variable is found to be using a different timezone, an error will be thrown.